### DISCUSSION TITLE
JavaScript Template String Expression Surround

### DISCUSSION THREAD

**Discussion by willdavidow** [kylechui/nvim-surround#266] [Q&A] [answered]
Hey all, I'm working on a custom surround/motion to quickly surround text within a string... 

So far I have this, and it works at the most-basic "proof of concept" level: 

```
surrounds = {
	["s"] = {
		add = { { "${" }, { "}" } },
	},
},
```

it works in visual mode and inner word motions and so on, with the `s` key as the modifier..

I'd like to make it "smarter," but I'm not 100% sure what I'm doing more advanced stuff within `nvim-surround` just yet.

Ideally I'm hoping to have something that functions like this:

If the nearest surrounding pair is a string that is either `'` or `"` I'd like to change that wrapper to a template literal string  surrounded by ` . for example: 

if my normal mode keymap is `yz` and I press `yziws` with my cursor over the word `string` I'd like  "some string of stuff" or 'some string of stuff' would change to \`some ${string} of stuff` 

If the nearest surrounding pair is not a quote of any kind, then the motion context should also create a template literal string around the proper context based on the motion: in normal mode if I type `yziws` with my cursor over the word random I'd like it to take (some random words somewhere) and change to (some `${random}` words somewhere) 

This last example I can think of would be a similar scenario as the previous example: if I have my cursor over the word `random` and enter a motion to surround all the text in the parenthesis I'd like it to go from (some random words somewhere) and change to (`some ${random} words somewhere`)

**Comment by kylechui** [comment, marked as answer]
There currently isn't a way to detect what the nearest surrounding pair is, and it seems quite complicated to expose this information to the end user as a part of the plugin itself. However, here are some potential workarounds:

- There's the `q` alias for targeting the nearest surrounding pair of quotes; your first case could be handled by first running `` csq` `` (change surrounding quotes to `` ` ``) then `yziws`.
- I might recommend you create a second surround key like `S` that `add`s ``{ "`${", "}`" }`` to the buffer to handle your second case

I think a *better*/marginally more complicated way is for you to use Tree-sitter to detect whether or not the node located at the cursor is a string, then use that information to selectively perform one of the two actions you mentioned above. I'll look into this more later, but don't currently have the time to experiment with it.

**Reply by willdavidow** [reply]
Oh that sounds like a good thread to pull on... I'll keep looking there.

I haven't checked yet, but would checking the source of this plugin for the way the `nearest surrounding pair of quotes` functionality works be a good example of creating a treesitter query?

Thanks for the help and suggestion(s).

**Reply by kylechui** [reply]
Ah no, the "nearest surrounding X" just finds all possible surrounds (using different "engines" like motions, pattern matching, TreeSitter, etc.) and chooses the nearest one. I  was recommending using Treesitter to determine if you're in a string, then act from there (shouldn't require queries, to my knowledge).

**Reply by willdavidow** [reply]
ahh, got it... alright, I'm starting to dig into the Treesitter documentation now and it doesn't seem too insane... Hopefully I'll have something cool to post in the showcase wiki soon! 

Thanks again for your help.

**Reply by willdavidow** [reply]
Apologies if this is a silly question: 

Is there a way to call csq`programatically? I've been trying to pluck a few things out from the plugin code, and while nothing is producing an error in what I'm trying, it's also not producing a result:

here's what I've been trying, it works in the context of knowing when i'm inside a quoted region, but none of the programmatic calls to `change_surround` are working

```
	surrounds = {
	["s"] = {
		add = function()
			-- local query = vim.treesitter
			local surround = require("nvim-surround")
			local config = require("nvim-surround.config")
			local ts_utils = require("nvim-treesitter.ts_utils")
			local cur = ts_utils.get_node_at_cursor(0, true)
			local language = vim.bo.filetype
			-- the lua options for not-long if statements seem wonky and overblown...
			local is_jsy = (
				language == "javascript"
				or language == "javascriptreact"
				or language == "typescript"
				or language == "typescriptreact"
			)

			if is_jsy then
				local cur_type = cur:type()
				if cur and (cur_type == "string" or cur_type == "string_fragment") then
					print("in a string; need to change to template " .. language)
					-- "q" being the alias for quotes
					surround.change_surround({
						curpos = cur,
						del_char = "q",
						char = "`",
						line_mode = false,
					})
					-- explicitly trying to delete single ' and replace with `
					surround.change_surround({
						curpos = cur,
						del_char = "'",
						char = "`",
						line_mode = false,
					})
					-- a table of the quotes i want to swap for `
					surround.change_surround({
						curpos = cur,
						del_char = { '"', "'" },
						char = "`",
						line_mode = false,
					})
					surround.change_surround({
						curpos = cur,
							del_char = { "q" },
							char = "`",
							line_mode = false,
						})
						-- neither of these work, but feel like i'm on the right track...
						--[[ return config.get_selection({
							char = "w",
							pattern = "^([\"'])().([\"'])()",
							replacement = { { "`" }, { "`" } },
						}) ]]
						--[[ surround.change_surround({
							curpos = cur,
							del_char = "q",
							add_delimeters = { { "`" }, { "`" } },
							line_mode = false,
						}) ]]
					elseif cur and cur_type == "template_string" then
						print("in template string; just surround with ${}" .. language)
					else
						print("not in a template string or a string; surround current word with ` and ${}  " .. language)
					end
				end
			end,
		},
	},
},
```

**Reply by kylechui** [reply]
I would try something like ``vim.cmd.normal("csq`")``.

**Reply by willdavidow** [reply]
Awesome - that worked!
